dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/sync/DataValuesSynchronization.java

```diff
@@ -0,0 +1,243 @@
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.HttpServerErrorException;
+import org.springframework.web.client.RequestCallback;
+import org.springframework.web.client.ResourceAccessException;
+import org.springframework.web.client.ResponseExtractor;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.Date;
+
+/**
+ * @author David Katuscak
+ */
+public class DataValuesSynchronization
+{
+    private static final Log log = LogFactory.getLog( DataValuesSynchronization.class );
+
+    private final DataValueService dataValueService;
+
+    private final DataValueSetService dataValueSetService;
+
+    private final SystemSettingManager systemSettingManager;
+
+    private final RestTemplate restTemplate;
+
+
+    @Autowired
+    public DataValuesSynchronization( DataValueService dataValueService, DataValueSetService dataValueSetService, SystemSettingManager systemSettingManager, RestTemplate restTemplate )
+    {
+        this.dataValueService = dataValueService;
+        this.dataValueSetService = dataValueSetService;
+        this.systemSettingManager = systemSettingManager;
+        this.restTemplate = restTemplate;
+    }
+
+    /**
+     * Gets the time of the last successful data synchronization operation. If not set,
+     * the current date subtracted by three days is returned.
+     */
+    private Date getLastDataSynchSuccessFallback()
+    {
+        return (Date) systemSettingManager.getSystemSetting( SettingKey.LAST_SUCCESSFUL_DATA_SYNC );
+    }
+
+    public SynchronizationResult syncDataValuesData()
+    {
+        try
+        {
+            if ( !SyncUtils.testServerAvailability( systemSettingManager, restTemplate ).isAvailable() )
+            {
+                return SynchronizationResult.newFailureResultWithMessage( "DataValues synchronization failed. Remote server is unavailable." );
+            }
+
+            log.info( "Starting DataValues synchronization job." );
+
+            // ---------------------------------------------------------------------
+            // Set time for last success to start of process to make data saved
+            // subsequently part of next synch process without being ignored
+            // ---------------------------------------------------------------------
+
+            final Date startTime = new Date();
+            final Date lastSuccessTime = getLastDataSynchSuccessFallback();
+            final int objectsToSync = dataValueService.getDataValueCountLastUpdatedAfter( lastSuccessTime, true );
+
+            if ( objectsToSync == 0 )
+            {
+                log.debug( "Skipping synchronization, no new or updated DataValues" );
+                return SynchronizationResult.newSuccessResultWithMessage( "Skipping synchronization, no new or updated DataValues" );
+            }
+
+            log.info( "Values: " + objectsToSync + " since last synchronization success: " + lastSuccessTime );
```

Comment:
Maybe write it like: "Found <objectsToSync> values since last synchronization (<lastSuccessTime>)